arc-theme: 20211018 -> 20220223

SebTM opened the pull request:
###### Description of changes

- Update to latest release
- Added FONTCONFIG_FILE to avoid "Fontconfig error: Cannot load default config file: No such file: (null)"
- Removed Parameter mesonFlag gtk3_version (upstream deleted)
- Cleanup (build all themes/variants)

###### Things done
`[ERROR] Fontconfig error: No writable cache directories`
This error occurs up to seven times per asset, I tired to fix it with [makeFontCache](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/fontconfig/make-fonts-cache.nix) but it seems to not help.

Then I came across that this is maybe not the point why the build is failing, but this should be:
`[ERROR] Running custom install script '/build/source/meson/install-file.py /build/source/build/common/cinnamon/cinnamon.css $MESON_INSTALL_DESTDIR_PREFIX/share/themes/Arc/cinnamon/cinnamon.css'`

I tried set `$MESON_INSTALL_DESTDIR_PREFIX `as I couldn't see it in the build summary but it does not work either.

Full build-log: [arc-build.log](https://github.com/NixOS/nixpkgs/files/8321867/arc-build.log)

Not directly related, but it made debugging harder: `nix log /nix/store/jgjlpxizwm0w29wx09v2y5c0gq5m75lg-arc-theme-20220223.drv` -> `error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override. Maybe there is another way to get the logs without having to enable experimental-features?

Do you have recommendations/tips on this?

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SebTM committed:
arc-theme: 20211018 -> 20220223

SuperSandro2000 commented:
> Not directly related, but it made debugging harder: nix log /nix/store/jgjlpxizwm0w29wx09v2y5c0gq5m75lg-arc-theme-20220223.drv -> `error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override. Maybe there is another way to get the logs without having to enable experimental-features?

IDK, I have them always enabled.

SuperSandro2000 commented on pkgs/data/themes/arc/default.nix:
```diff
@@ -45,12 +46,14 @@ stdenv.mkDerivation rec {
   '';
 
+  # Fontconfig error: Cannot load default config file: No such file: (null)
+  FONTCONFIG_FILE="${fontconfig.out}/etc/fonts/fonts.conf";
```
```suggestion
  FONTCONFIG_FILE = "${fontconfig.out}/etc/fonts/fonts.conf";
```

SuperSandro2000 commented on pkgs/data/themes/arc/default.nix:
```diff
@@ -11,17 +11,18 @@
 , optipng
 , inkscape
 , cinnamon
+, fontconfig
```
```suggestion
, makeFontsConf
```

SuperSandro2000 commented:
Random guess: It has something to do with the custom install file https://github.com/jnsh/arc-theme/blob/master/meson/install-file.py#L36-L40 because it only fails when using remote builder, not when adding ``--builders ""``.

SebTM commented:
> Random guess: It has something to do with the custom install file https://github.com/jnsh/arc-theme/blob/master/meson/install-file.py#L36-L40 because it only fails when using remote builder, not when adding `--builders ""`.

Hmm, seems to be implemented after the last release we had packaged. I will try to ask upstream for help...

jnsh commented:
Besides the build error: `pkg-config`, `optipng` and `gtk3` aren't required as build dependencies any more.

Also you can avoid the `inkscape`dependency by building from a release tarball instead of fetching the code from git, as I mentioned [here](https://github.com/jnsh/arc-theme/issues/178#issuecomment-1076554848). Although, I'm not sure whether that's feasible or smart with nixpkgs :)

jtojnar commented:
Build sandbox does not have `/usr/bin/env` or `python3`, you need to patch the script:

```patch
diff --git a/pkgs/data/themes/arc/default.nix b/pkgs/data/themes/arc/default.nix
index ae3c333ab27..78c1bb2cfef 100644
--- a/pkgs/data/themes/arc/default.nix
+++ b/pkgs/data/themes/arc/default.nix
@@ -3,6 +3,7 @@
 , sassc
 , meson
 , ninja
+, python3
 , pkg-config
 , gtk3
 , glib
@@ -34,6 +35,7 @@ stdenv.mkDerivation rec {
     inkscape
     gtk3
     glib # for glib-compile-resources
+    python3
   ];
 
   propagatedUserEnvPkgs = [
@@ -41,6 +43,10 @@ stdenv.mkDerivation rec {
     gtk-engine-murrine
   ];
 
+  postPatch = ''
+    patchShebangs meson/install-file.py
+  '';
+
   preBuild = ''
     # Shut up inkscape's warnings about creating profile directory
     export HOME="$NIX_BUILD_ROOT"
```

jtojnar commented on pkgs/data/themes/arc/default.nix:
```diff
@@ -45,12 +46,14 @@ stdenv.mkDerivation rec {
     export HOME="$NIX_BUILD_ROOT"
```
That is an empty variable, use `$TMPDIR`.

jnsh commented:
> Build sandbox does not have `/usr/bin/env` or `python3`, you need to patch the script:

I may be missing something about your build system, but wouldn't `python3` be already present in the build environment as a dependency for `meson`?

SebTM commented:
Thanks so much for your help! I've adjusted the PR to all of your hints, switched to mkFontConf, removed unneeded build dependencies, python fix.

The build-env python "issue" makes so much sense as you explained it sad that I didn't came across it before :/

-----

> > Build sandbox does not have `/usr/bin/env` or `python3`, you need to patch the script:
> 
> I may be missing something about your build system, but wouldn't `python3` be already present in the build environment as a dependency for `meson`?

I guess it has to do with the build-system isolation nix does but I can't explain it well like others here can ✌🏻

romildo commented:
Why is `ugrep` being updated in this PR? Is it related to `arc-theme`?

jtojnar commented:
> I may be missing something about your build system, but wouldn't `python3` be already present in the build environment as a dependency for `meson`?

Nixpkgs’s generic builder will only make available in the build environment:

- direct dependencies (listed e.g. in `buildInputs`)
- dependencies explicitly propagated by dependencies (even transitively).

Since Meson being implemented in Python is an implementation detail, we do not propagate Python to dependents of Meson package.

SebTM commented:
> Why is `ugrep` being updated in this PR? Is it related to `arc-theme`?

It hasn't sorry, my fault - fixed 👍🏻
